<a href="https://colab.research.google.com/github/DeboraOliver/MachineLearning_lessons/blob/main/E0_3_Redes_Neurais.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Estudo Orientado 3
Você deve criar um modelo de predição usando Redes Neurais do tipo perceptron. Além disso, deve alterar os parâmetros de treino do modelo e analisar o comportamento do modelo gerado, avaliando o aumento ou redução da acurácia.

Primeiro, você deve alterar a quantidade de iterações de treino da rede neural e perceber se a acurácia do modelo gerado aumentou ou diminuiu.

Após isso, você deve alterar a quantidade de camadas escondidas e de neurônios em cada camada, novamente, avaliando se a acurácia do modelo aumentou ou diminuiu.

Para criar o modelo, use o código a seguir e então faça as modificações propostas, descritas em detalhe a seguir.

 

Altere a quantidade de iterações para 100, 500, 1000, 2000 e 5000.<br>

 Avalie os resultados de acurácia para cada quantidade de iterações.
Para os testes a seguir, mantenha a quantidade de iterações em 1000 e faça as alterações propostas:<br>

Altere a quantidade de camadas escondidas para 1 com 5 neurônios.<br>
Altere a quantidade de camadas escondidas para 3 com 5 neurônios. <br>
Altere a quantidade de camadas escondidas para 1 com 10 neurônios.<br>
Altere a quantidade de camadas escondidas para 3 com 10 neurônios.<br>
Altere a quantidade de camadas escondidas para 5 com 50 neurônios.

In [4]:
import pandas as pd

from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from sklearn.neural_network import MLPClassifier

import warnings
warnings.filterwarnings("ignore")

In [5]:
# fAZ O DOWNLOAD DOS DADOS E SALVA NO DATAFRAME
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
names=['sepal-length','sepal-width','petal-length', 'petal-width','Classe']
irisdata = pd.read_csv(url, names = names)

def redes_neurais(irisdata, camadas_escondidas, max_iteracoes):

  X_caracteristicas = irisdata.iloc[:,0:4]
  Y_classes = irisdata.iloc[:,4]

  le = preprocessing.LabelEncoder()
  y = le.fit_transform(Y_classes)

  X_train, X_test, y_train, y_test = train_test_split(X_caracteristicas, y, test_size = 0.3, random_state= 0)

  #cria o modelo de  classificacao usando a rede neural com 3 camadas escondidas e 10 perceptrons
  mlp = MLPClassifier(hidden_layer_sizes=camadas_escondidas, max_iter = max_iteracoes)

  mlp.fit(X_train, y_train)

  predictions = mlp.predict(X_test)

  acuracia = accuracy_score(y_test, predictions)

  return print("Acurácia: ",acuracia)

# Alterando a quantidade de iterações para 100, 500, 1000, 2000 e 5000.

In [6]:
redes_neurais(irisdata, camadas_escondidas = (10,10,10), max_iteracoes=100)

Acurácia:  0.6888888888888889


In [7]:
redes_neurais(irisdata, camadas_escondidas = (10,10,10), max_iteracoes=500)

Acurácia:  0.9777777777777777


In [8]:
redes_neurais(irisdata, camadas_escondidas = (10,10,10), max_iteracoes=1000)

Acurácia:  0.9777777777777777


In [9]:
redes_neurais(irisdata, camadas_escondidas = (10,10,10), max_iteracoes=2000)

Acurácia:  0.9777777777777777


In [10]:
redes_neurais(irisdata, camadas_escondidas = (10,10,10), max_iteracoes=5000)

Acurácia:  0.9777777777777777


# Altere a quantidade de camadas escondidas para 1 com 5 neurônios.


In [11]:
redes_neurais(irisdata, camadas_escondidas = (5), max_iteracoes=1000)

Acurácia:  0.6


# Altere a quantidade de camadas escondidas para 3 com 5 neurônios.


In [12]:
redes_neurais(irisdata, camadas_escondidas = (5,5,5), max_iteracoes=1000)

Acurácia:  0.6


# Altere a quantidade de camadas escondidas para 1 com 10 neurônios.


In [13]:
redes_neurais(irisdata, camadas_escondidas = (10), max_iteracoes=1000)

Acurácia:  0.7555555555555555


# Altere a quantidade de camadas escondidas para 3 com 10 neurônios.


In [14]:
redes_neurais(irisdata, camadas_escondidas = (10,10,10), max_iteracoes=1000)

Acurácia:  0.9777777777777777


# Altere a quantidade de camadas escondidas para 5 com 50 neurônios

In [15]:
redes_neurais(irisdata, camadas_escondidas = (50,50,50,50,50), max_iteracoes=1000)

Acurácia:  0.9777777777777777


# Conclusão

O objetivo é um valor de acurácia mais alto possível. Isto é, quanto maior for o valor dessa métrica, melhor é o modelo. Portanto, no exercício acima a melhor acurácia foi obtida quando aumentamos o número de iterações de 100 para 500, onde a acurácia chega a aproximadamente 98%. Nota-se que quando aumentamos a quantidade de iterações (acima de 500) o valor da acurácia não aumenta mais. Por outro lado, aumentando a quantidade de camadas e de neurônios, o maior valor de acurácia é atingido quando usamos 3 camadas escondidas com 10 neurônios. Ao aumentar a quantidade de camadas escondidas para 5 e o número de neurônios para 50, não há alteração da acurácia, pois a técnica já está chegando próximo ao melhor valor possível.